File: packages/common-tools/project-tools.ts
import { z } from "zod";
import { exists } from "@std/fs";
import { join } from "@std/path";
import type { MCPTool } from "@packages/mcp-core/types.ts";

// Schemas
export const FindProjectsInputSchema = z.object({
  root_directory: z.string().optional().describe("Root directory to search for projects"),
  project_type: z.string().optional().describe("Type of projects to find (orchestration, npm, deno, etc.)"),
});

export const FindProjectsOutputSchema = z.object({
  projects: z.array(z.object({
    name: z.string(),
    path: z.string(),
    type: z.string(),
    metadata: z.record(z.any()).optional(),
  })),
  total: z.number(),
});

export const GetProjectInfoInputSchema = z.object({
  project_path: z.string().describe("Path to the project"),
});

export const GetProjectInfoOutputSchema = z.object({
  name: z.string(),
  path: z.string(),
  type: z.string(),
  config_files: z.array(z.string()),
  metadata: z.record(z.any()).optional(),
});

export const CreateProjectInputSchema = z.object({
  name: z.string().describe("Name of the project"),
  path: z.string().describe("Path where to create the project"),
  template: z.string().optional().describe("Project template to use"),
});

export const CreateProjectOutputSchema = z.object({
  success: z.boolean(),
  name: z.string(),
  path: z.string(),
  files_created: z.array(z.string()),
});

// Tool definitions
export const projectTools: MCPTool[] = [
  {
    name: "find_projects",
    description: "Find projects in a directory tree",
    inputSchema: FindProjectsInputSchema,
    outputSchema: FindProjectsOutputSchema,
  },
  {
    name: "get_project_info",
    description: "Get detailed information about a project",
    inputSchema: GetProjectInfoInputSchema,
    outputSchema: GetProjectInfoOutputSchema,
  },
  {
    name: "create_project",
    description: "Create a new project from template",
    inputSchema: CreateProjectInputSchema,
    outputSchema: CreateProjectOutputSchema,
  },
];

// Project type detectors
const projectDetectors = {
  orchestration: (path: string) => exists(join(path, ".orchestration.json")),
  npm: (path: string) => exists(join(path, "package.json")),
  deno: (path: string) => exists(join(path, "deno.json")),
  cargo: (path: string) => exists(join(path, "Cargo.toml")),
  python: async (path: string) => {
    return await exists(join(path, "pyproject.toml")) ||
           await exists(join(path, "requirements.txt")) ||
           await exists(join(path, "setup.py"));
  },
  git: (path: string) => exists(join(path, ".git")),
};

// Tool handlers
export const projectHandlers = {
  find_projects: async (args: z.infer<typeof FindProjectsInputSchema>) => {
    const { root_directory = Deno.cwd(), project_type } = args;
    
    try {
      const projects = [];
      
      async function scanDirectory(dirPath: string, depth = 0) {
        if (depth > 3) return; // Limit recursion depth
        
        try {
          for await (const entry of Deno.readDir(dirPath)) {
            if (entry.isDirectory && !entry.name.startsWith(".")) {
              const entryPath = join(dirPath, entry.name);
              
              // Check each project type
              for (const [type, detector] of Object.entries(projectDetectors)) {
                if (project_type && type !== project_type) continue;
                
                if (await detector(entryPath)) {
                  let metadata;
                  
                  // Try to load project metadata
                  try {
                    if (type === "orchestration") {
                      const metaPath = join(entryPath, ".orchestration.json");
                      metadata = JSON.parse(await Deno.readTextFile(metaPath));
                    } else if (type === "npm") {
                      const pkgPath = join(entryPath, "package.json");
                      metadata = JSON.parse(await Deno.readTextFile(pkgPath));
                    } else if (type === "deno") {
                      const denoPath = join(entryPath, "deno.json");
                      metadata = JSON.parse(await Deno.readTextFile(denoPath));
                    }
                  } catch {
                    // Ignore metadata errors
                  }
                  
                  projects.push({
                    name: entry.name,
                    path: entryPath,
                    type,
                    metadata,
                  });
                  break; // Don't check other types for this directory
                }
              }
              
              // Recurse into subdirectories
              await scanDirectory(entryPath, depth + 1);
            }
          }
        } catch {
          // Ignore directories we can't read
        }
      }
      
      await scanDirectory(root_directory);
      
      return {
        projects,
        total: projects.length,
      };
    } catch (error) {
      throw new Error(`Failed to find projects: ${error instanceof Error ? error.message : String(error)}`);
    }
  },

  get_project_info: async (args: z.infer<typeof GetProjectInfoInputSchema>) => {
    const { project_path } = args;
    
    try {
      if (!await exists(project_path)) {
        throw new Error(`Project path ${project_path} does not exist`);
      }
      
      // Determine project type
      let projectType = "unknown";
      let metadata;
      const configFiles = [];
      
      for (const [type, detector] of Object.entries(projectDetectors)) {
        if (await detector(project_path)) {
          projectType = type;
          break;
        }
      }
      
      // Collect config files
      const commonConfigFiles = [
        ".orchestration.json",
        "package.json",
        "deno.json",
        "Cargo.toml",
        "pyproject.toml",
        "requirements.txt",
        "setup.py",
        ".gitignore",
        "README.md",
      ];
      
      for (const file of commonConfigFiles) {
        if (await exists(join(project_path, file))) {
          configFiles.push(file);
        }
      }
      
      // Load metadata based on project type
      try {
        if (projectType === "orchestration") {
          const metaPath = join(project_path, ".orchestration.json");
          metadata = JSON.parse(await Deno.readTextFile(metaPath));
        } else if (projectType === "npm") {
          const pkgPath = join(project_path, "package.json");
          metadata = JSON.parse(await Deno.readTextFile(pkgPath));
        } else if (projectType === "deno") {
          const denoPath = join(project_path, "deno.json");
          metadata = JSON.parse(await Deno.readTextFile(denoPath));
        }
      } catch {
        // Ignore metadata errors
      }
      
      const projectName = metadata?.name || project_path.split("/").pop() || "unknown";
      
      return {
        name: projectName,
        path: project_path,
        type: projectType,
        config_files: configFiles,
        metadata,
      };
    } catch (error) {
      throw new Error(`Failed to get project info: ${error instanceof Error ? error.message : String(error)}`);
    }
  },

  create_project: async (args: z.infer<typeof CreateProjectInputSchema>) => {
    const { name, path, template = "basic" } = args;
    
    try {
      const projectPath = join(path, name);
      
      if (await exists(projectPath)) {
        throw new Error(`Project directory ${projectPath} already exists`);
      }
      
      await Deno.mkdir(projectPath, { recursive: true });
      const filesCreated = [projectPath];
      
      // Create basic project structure based on template
      if (template === "deno" || template === "basic") {
        const denoConfig = {
          name,
          version: "1.0.0",
          tasks: {
            start: "deno run --allow-all main.ts",
            test: "deno test",
            fmt: "deno fmt",
            lint: "deno lint",
          },
        };
        
        await Deno.writeTextFile(
          join(projectPath, "deno.json"),
          JSON.stringify(denoConfig, null, 2)
        );
        filesCreated.push("deno.json");
        
        await Deno.writeTextFile(
          join(projectPath, "main.ts"),
          `// ${name} - Main entry point\n\nfunction main() {\n  console.log("Hello from ${name}!");\n}\n\nif (import.meta.main) {\n  main();\n}\n`
        );
        filesCreated.push("main.ts");
      }
      
      // Create README
      await Deno.writeTextFile(
        join(projectPath, "README.md"),
        `# ${name}\n\nA new project created with common-tools.\n\n## Getting Started\n\n\`\`\`bash\ncd ${name}\n# Add your instructions here\n\`\`\`\n`
      );
      filesCreated.push("README.md");
      
      return {
        success: true,
        name,
        path: projectPath,
        files_created: filesCreated,
      };
    } catch (error) {
      throw new Error(`Failed to create project: ${error instanceof Error ? error.message : String(error)}`);
    }
  },
};

File: python-packages/common_tools/project_tools.py
"""
Project discovery and management tools
"""
import json
from pathlib import Path
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field

from mcp_core.types import MCPTool


# Schema definitions
class ProjectMetadata(BaseModel):
    """Project metadata"""
    name: str
    type: str
    config_files: List[str]
    metadata: Optional[Dict[str, Any]] = None


class FindProjectsInputSchema(BaseModel):
    """Schema for find projects input"""
    root_directory: Optional[str] = Field(default=".", description="Root directory to search for projects")
    project_type: Optional[str] = Field(default=None, description="Type of projects to find")


class FindProjectsOutputSchema(BaseModel):
    """Schema for find projects output"""
    projects: List[ProjectMetadata]
    total: int


class GetProjectInfoInputSchema(BaseModel):
    """Schema for get project info input"""
    project_path: str = Field(description="Path to the project")


class GetProjectInfoOutputSchema(BaseModel):
    """Schema for get project info output"""
    name: str
    path: str
    type: str
    config_files: List[str]
    metadata: Optional[Dict[str, Any]] = None


class CreateProjectInputSchema(BaseModel):
    """Schema for create project input"""
    name: str = Field(description="Project name")
    path: str = Field(description="Project path")
    project_type: Optional[str] = Field(default="basic", description="Project type")


class CreateProjectOutputSchema(BaseModel):
    """Schema for create project output"""
    success: bool
    name: str
    path: str


# Tool definitions
project_tools: List[MCPTool] = [
    MCPTool(
        name="find_projects",
        description="Find projects in a directory",
        input_schema=FindProjectsInputSchema.model_json_schema(),
        output_schema=FindProjectsOutputSchema.model_json_schema(),
    ),
    MCPTool(
        name="get_project_info",
        description="Get information about a specific project",
        input_schema=GetProjectInfoInputSchema.model_json_schema(),
        output_schema=GetProjectInfoOutputSchema.model_json_schema(),
    ),
    MCPTool(
        name="create_project",
        description="Create a new project",
        input_schema=CreateProjectInputSchema.model_json_schema(),
        output_schema=CreateProjectOutputSchema.model_json_schema(),
    ),
]


# Helper functions
def detect_project_type(path: Path) -> tuple[str, List[str]]:
    """Detect project type based on config files"""
    config_files = []
    project_type = "unknown"
    
    # Check for common project files
    if (path / "package.json").exists():
        config_files.append("package.json")
        project_type = "nodejs"
    
    if (path / "deno.json").exists():
        config_files.append("deno.json")
        project_type = "deno"
    
    if (path / "pyproject.toml").exists():
        config_files.append("pyproject.toml")
        project_type = "python"
    
    if (path / "Cargo.toml").exists():
        config_files.append("Cargo.toml")
        project_type = "rust"
    
    if (path / ".orchestration.json").exists():
        config_files.append(".orchestration.json")
        project_type = "orchestration"
    
    return project_type, config_files


# Tool handlers
async def find_projects_handler(args: dict) -> dict:
    """Find projects handler"""
    validated_args = FindProjectsInputSchema(**args)
    root_path = Path(validated_args.root_directory)
    
    try:
        projects = []
        
        # Search for projects recursively
        for item in root_path.rglob("*"):
            if item.is_dir():
                project_type, config_files = detect_project_type(item)
                
                # Filter by project type if specified
                if validated_args.project_type and project_type != validated_args.project_type:
                    continue
                
                if config_files:  # Only include if we found config files
                    metadata = None
                    
                    # Try to extract metadata from config files
                    if "package.json" in config_files:
                        try:
                            with open(item / "package.json") as f:
                                pkg_data = json.load(f)
                                metadata = {
                                    "name": pkg_data.get("name"),
                                    "version": pkg_data.get("version"),
                                    "description": pkg_data.get("description")
                                }
                        except:
                            pass
                    
                    projects.append(ProjectMetadata(
                        name=item.name,
                        type=project_type,
                        config_files=config_files,
                        metadata=metadata
                    ))
        
        return FindProjectsOutputSchema(
            projects=projects,
            total=len(projects)
        ).model_dump()
        
    except Exception as error:
        raise Exception(f"Failed to find projects in {root_path}: {str(error)}")


async def get_project_info_handler(args: dict) -> dict:
    """Get project info handler"""
    validated_args = GetProjectInfoInputSchema(**args)
    project_path = Path(validated_args.project_path)
    
    try:
        if not project_path.exists():
            raise Exception(f"Project path {project_path} does not exist")
        
        project_type, config_files = detect_project_type(project_path)
        
        metadata = None
        if "package.json" in config_files:
            try:
                with open(project_path / "package.json") as f:
                    pkg_data = json.load(f)
                    metadata = {
                        "name": pkg_data.get("name"),
                        "version": pkg_data.get("version"),
                        "description": pkg_data.get("description")
                    }
            except:
                pass
        
        return GetProjectInfoOutputSchema(
            name=project_path.name,
            path=str(project_path),
            type=project_type,
            config_files=config_files,
            metadata=metadata
        ).model_dump()
        
    except Exception as error:
        raise Exception(f"Failed to get project info for {project_path}: {str(error)}")


async def create_project_handler(args: dict) -> dict:
    """Create project handler"""
    validated_args = CreateProjectInputSchema(**args)
    project_path = Path(validated_args.path)
    
    try:
        # Create project directory
        project_path.mkdir(parents=True, exist_ok=True)
        
        # Create basic project structure based on type
        if validated_args.project_type == "python":
            # Create basic Python project structure
            (project_path / "src").mkdir(exist_ok=True)
            (project_path / "tests").mkdir(exist_ok=True)
            
            # Create pyproject.toml
            pyproject_content = f"""[project]
name = "{validated_args.name}"
version = "0.1.0"
description = ""
requires-python = ">=3.11"
dependencies = []

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"
"""
            with open(project_path / "pyproject.toml", "w") as f:
                f.write(pyproject_content)
        
        return CreateProjectOutputSchema(
            success=True,
            name=validated_args.name,
            path=str(project_path)
        ).model_dump()
        
    except Exception as error:
        raise Exception(f"Failed to create project {validated_args.name}: {str(error)}")


# Handler mapping
project_handlers = {
    "find_projects": find_projects_handler,
    "get_project_info": get_project_info_handler,
    "create_project": create_project_handler,
}